test/modules/DirectMethodSender/DirectMethodCloudSender.cs

```diff
@@ -45,7 +45,12 @@ sealed class DirectMethodCloudSender : DirectMethodSenderBase
             ulong directMethodCount,
             CancellationToken none)
         {
+            // serviceClientTimeouts needs to be explicitly set to let IoTHub figure out where does the device is connected to the IoTHub server.
```

Comment:
* Does it mean device can't figure out what iot hub server to connect to? *
Nope. It's quite the opposite actually. Say we have a deviceX that was connected to IoTHub serverA... then a network disconnected, the deviceX reconnected to IoTHub serverB. Now we have a ServiceClient that tries to send a directMethod to the deviceX, because of caching in IoTHub it would thinks that deviceX is still connected to serverA -- this would immediately throw DeviceNotFoundException right away by default. By setting a timeout on the ServiceClient, it would actually allow IoTHub to resolve that deviceX is no longer connected to serverA, and reroute the directMethod to serverB. 

* Do you have any reference for clarity? *
Not really. David and I looked into ServiceClient code on how it handles the invokeMethod request. and combine w/ David's background knowledge on IoTHub backend.

Reply:
Based what you mentioned above, does it mean setting connection timeout will make IoT Hub to clean up their cache faster?  And this sounds like an IoT hub problem, you may need to check logs from Kusto or ask IoT Hub team for confirmation of your understandings.

Would you please share what line of code in ServiceClient make you think ConnectionTime will fix the problem?

Reply:
From the DirectMethodSender log we got ` "Microsoft.Azure.Devices.Common.Exceptions.DeviceNotFoundException: Device {\"Message\":\"{"errorCode":404103,"trackingId":"8503982a657245948d4d6e071ae25e7c-TimeStamp:05/21/2020 01:02:16","message":"Timed out waiting for device to connect.","info":{"timeout":"00:00:00"},"timestampUtc":"2020-05-21T01:02:16.2521655Z"}\",\"ExceptionMessage\":\"\"} not registered", `
Note the `{"timeout":"00:00:00"}`

This is where the timeout is calculated 
https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/service/src/AmqpServiceClient.cs#L296-L305

Reply:
I am confused.  Before your chage, CloudToDeviceMethod doesn't pass ConnectionTimeout and default value is zero. From the logic you referred, Direct method operation timeout doesn't change as there is a lower bound logic set to 100 seconds.  Do you want to increase or decrease timeout?

Reply:
Increase timeout. Looks like I forgot in include the 100s in the calculation. Let me make it 110s.